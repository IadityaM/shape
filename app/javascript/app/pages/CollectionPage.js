import React, { Component } from 'react'
import { Link } from 'react-router-dom'
import { inject, observer } from 'mobx-react'

import withAuth from '~/utils/withAuth'

const fakeCollections = [
  { id: 1, name: 'x' },
  { id: 2, name: 'y' },
  { id: 3, name: 'z' },
  { id: 4, name: 'zzzz' },
]

const CollectionsView = ({ collections }) => (
  collections.map(c => (
    <div key={c.id}>
      <Link to={`/collections/${c.id}`}>
        {c.name}
      </Link>
      <hr />
    </div>
  ))
)

@withAuth
@inject('routingStore', 'authStore')
@observer
class CollectionPage extends Component {
  doStuff = () => {
    this.props.authStore.loadUser({
      id: 1,
      firstName: 'Mister',
      lastName: 'Tester',
      email: 'mister@tester.com',
    })
  }

  render () {
    const { routingStore, authStore } = this.props
    const testThing = authStore.currentUser ? authStore.currentUser.fullName() : 'nope.'
    return (
      <div>
        <h1>Collection Page</h1>
        <div>
          { routingStore.location.pathname }
        </div>
        <button onClick={this.doStuff}>Heyo</button>
        <div>
          { JSON.stringify(authStore) }
          { testThing }
        </div>

        <CollectionsView collections={fakeCollections} />
      </div>
    )
  }
}

export default CollectionPage
