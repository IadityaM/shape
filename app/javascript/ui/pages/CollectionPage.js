import { Fragment } from 'react'
import pluralize from 'pluralize'
import ReactRouterPropTypes from 'react-router-prop-types'
import { observable, action } from 'mobx'
import { inject, observer, PropTypes as MobxPropTypes } from 'mobx-react'

import PageError from '~/ui/global/PageError'
import PageWithApi from '~/ui/pages/PageWithApi'
import Loader from '~/ui/layout/Loader'
import PageContainer from '~/ui/layout/PageContainer'
import PageSeparator from '~/ui/global/PageSeparator'
import CollectionGrid from '~/ui/grid/CollectionGrid'
import MoveModal from '~/ui/grid/MoveModal'
import PageHeader from '~/ui/pages/shared/PageHeader'
import SubmissionBoxSettingsModal from '~/ui/submission_box/SubmissionBoxSettingsModal'

const isHomepage = ({ params }) => (params.org && !params.id)

@inject('apiStore', 'uiStore', 'routingStore')
@observer
class CollectionPage extends PageWithApi {
  @observable loadingSubmissions = false

  componentWillReceiveProps(nextProps) {
    super.componentWillReceiveProps(nextProps)
    // when navigating between collections, close BCT
    if (nextProps.match.params.id !== this.props.match.params.id) {
      this.props.uiStore.closeBlankContentTool()
    }
  }

  @action setLoadingSubmissions = val => {
    const { submissions_collection } = this.collection
    if (submissions_collection && submissions_collection.cardIds.length) {
      // if submissions_collection is preloaded with some cards, no need to show loader
      this.loadingSubmissions = false
      return
    }
    this.loadingSubmissions = val
  }

  get isHomepage() {
    return isHomepage(this.props.match)
  }

  get collection() {
    const { match, apiStore } = this.props
    if (!apiStore.collections.length) return null
    if (this.isHomepage) {
      return apiStore.find('collections', apiStore.currentUser.current_user_collection_id)
    }
    return apiStore.find('collections', match.params.id)
  }

  get roles() {
    const { apiStore, match } = this.props
    return apiStore.findAll('roles').filter((role) =>
      role.resource && role.resource.id === parseInt(match.params.id))
  }

  requestPath = (props) => {
    const { match, apiStore } = props
    if (isHomepage(match)) {
      return `collections/${apiStore.currentUser.current_user_collection_id}`
    }
    return `collections/${match.params.id}`
  }

  onAPILoad = async (response) => {
    this.updateError(null)
    const collection = response.data
    const { apiStore, uiStore, location } = this.props
    uiStore.setViewingCollection(collection)
    // setViewingCollection has to happen first bc we use it in openBlankContentTool
    if (!collection.collection_cards.length) {
      uiStore.openBlankContentTool()
    }
    collection.checkCurrentOrg()
    if (collection.isNormalCollection) {
      const thread = await apiStore.findOrBuildCommentThread(collection)
      if (location.search) {
        const menu = uiStore.openOptionalMenus(location.search)
        if (menu === 'comments') {
          uiStore.expandThread(thread.key)
        }
      }
      if (collection.isSubmissionBox && collection.submissions_collection) {
        this.setLoadingSubmissions(true)
        await apiStore.fetch('collections', collection.submissions_collection.id, true)
        this.setLoadingSubmissions(false)
      }
    } else {
      apiStore.clearUnpersistedThreads()
    }
  }

  updateCollection = () => {
    // TODO: what if there's no collection?
    // calling .save() will receive any API updates and sync them
    this.collection.API_updateCards()
    const { uiStore } = this.props
    uiStore.trackEvent('update', this.collection)
  }

  updateCollectionName = (name) => {
    this.collection.name = name
    this.collection.save()
    const { uiStore } = this.props
    uiStore.trackEvent('update', this.collection)
  }

  get submissionsPageSeparator() {
    const { collection } = this
    const { submissionTypeName } = collection
    return (
      <PageSeparator title={(
        <h3>
          {collection.submissions_collection.collection_cards.length}
          {' '}
          {collection.submissions_collection.collection_cards.length === 1
            ? submissionTypeName
            : pluralize(submissionTypeName)
          }
        </h3>
      )}
      />
    )
  }

  render() {
    // this.error comes from PageWithApi
    if (this.error) return <PageError error={this.error} />
    const { collection } = this
    // for some reason collection can come through as an object, but not some fields like can_edit,
    // which indicates it hasn't finished loading everything
    if (!collection || collection.can_edit === undefined) return <Loader />

    const { uiStore } = this.props
    // submissions_collection will only exist for submission boxes
    const { submissions_collection } = collection
    const {
      blankContentToolState,
      submissionBoxSettingsOpen,
      gridSettings,
    } = uiStore
    const { movingCardIds, cardAction } = uiStore
    // only tell the Grid to hide "movingCards" if we're moving and not linking
    const uiMovingCardIds = cardAction === 'move' ? movingCardIds : []
    // SharedCollection has special behavior where it sorts by most recently updated
    const sortBy = collection.isSharedCollection ? 'updated_at' : 'order'

    return (
      <Fragment>
        <PageHeader
          record={collection}
          isHomepage={this.isHomepage}
        />
        <PageContainer>
          <CollectionGrid
            // pull in cols, gridW, gridH, gutter
            {...gridSettings}
            gridSettings={gridSettings}
            updateCollection={this.updateCollection}
            collection={collection}
            canEditCollection={collection.can_edit_content}
            // Pass in cardIds so grid will re-render when they change
            cardIds={collection.cardIds}
            // Pass in BCT state so grid will re-render when open/closed
            blankContentToolState={blankContentToolState}
            movingCardIds={uiMovingCardIds}
            // passing length prop seems to properly trigger a re-render
            movingCards={uiStore.movingCardIds.length}
            sortBy={sortBy}
            // don't add the extra row for submission box
            addEmptyCard={!collection.isSubmissionBox}
          />
          {(collection.requiresSubmissionBoxSettings || submissionBoxSettingsOpen) &&
            <SubmissionBoxSettingsModal
              collection={collection}
            />
          }
          <MoveModal />
          { submissions_collection && (
            <div>
              { this.loadingSubmissions
                ? <Loader />
                : (
                  <div>
                    {this.submissionsPageSeparator}
                    <CollectionGrid
                      {...gridSettings}
                      updateCollection={this.updateCollection}
                      collection={submissions_collection}
                      canEditCollection={false}
                      // Pass in cardIds so grid will re-render when they change
                      cardIds={submissions_collection.cardIds}
                      // Pass in BCT state so grid will re-render when open/closed
                      blankContentToolState={
                        blankContentToolState.collectionId === submissions_collection.id
                          ? blankContentToolState
                          : {}
                      }
                      submissionSettings={{
                        type: collection.submission_box_type,
                        template: collection.submission_template,
                      }}
                      movingCardIds={[]}
                      movingCards={false}
                      sortBy={sortBy}
                    />
                  </div>
                )}
            </div>
          )}
        </PageContainer>
      </Fragment>
    )
  }
}

CollectionPage.propTypes = {
  match: ReactRouterPropTypes.match.isRequired,
  location: ReactRouterPropTypes.location.isRequired,
}
CollectionPage.wrappedComponent.propTypes = {
  apiStore: MobxPropTypes.objectOrObservableObject.isRequired,
  uiStore: MobxPropTypes.objectOrObservableObject.isRequired,
  routingStore: MobxPropTypes.objectOrObservableObject.isRequired,
}

export default CollectionPage
