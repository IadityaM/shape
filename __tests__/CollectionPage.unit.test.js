import CollectionPage from '~/ui/pages/CollectionPage'
import createComponentWithIntl from '../js_test_config/intl-enzyme-test-helper.js'

const id = 1
const collections = [
  { id: 1, name: 'My Workspace X' }
]
let wrapper, match, request, find, apiStore
let props
beforeEach(() => {
  match = { params: { id } }
  request = jest.fn()
  request.mockReturnValue(Promise.resolve({}))
  find = jest.fn()
  find.mockReturnValue(collections[0])
  apiStore = {
    request,
    find,
    sync: jest.fn(),
    collections,
  }
  props = { apiStore, match }
  wrapper = shallow(
    <CollectionPage.wrappedComponent {...props} />
  )
})

describe('CollectionPage', () => {
  it('makes an API call to fetch the collection', () => {
    expect(request).toBeCalledWith(`collections/${match.params.id}`)
  })

  it('displays the collection name', () => {
    expect(wrapper.find('h1').at(0).text()).toBe('Collection: My Workspace X')
  })

  it('renders correctly', () => {
    wrapper = createComponentWithIntl(
      <CollectionPage.wrappedComponent {...props} />
    )
    const tree = wrapper.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
