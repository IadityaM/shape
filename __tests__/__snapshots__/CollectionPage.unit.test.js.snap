// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollectionPage renders correctly 1`] = `
.c1 {
  font-size: 0.75rem;
  font-family: 'Gotham';
  text-transform: uppercase;
}

.c0 {
  z-index: 9999;
  position: fixed;
  top: 0;
  width: calc(100% - 4rem);
  background: #f2f2f2;
  padding: 1rem 2rem;
}

.c2 {
  margin-top: 90px;
  padding: 0 2rem;
}

<div>
  <header
    className="c0"
  >
    <div
      className="_Rfx0_ _Rfx1_ _Rfx2_"
    >
      <div
        className=""
      >
        <div
          className="c1"
        >
          <span>
            Innovation OS
          </span>
        </div>
      </div>
      <div
        className=""
      >
        <div
          className="MuiAvatar-root-2 Header-smallAvatar-1"
        >
          <img
            alt={undefined}
            className="MuiAvatar-img-4"
            sizes={undefined}
            src="https://picsum.photos/38"
            srcSet={undefined}
          />
        </div>
      </div>
    </div>
  </header>
  <main
    className="c2"
  >
    <h1>
      Collection: 
      My Workspace X
    </h1>
    <div
      className="Grid"
    />
  </main>
</div>
`;
